### PR TITLE
overlord/snapshotstate: store the SnapID in snapshot, block restore if changed

### PR DESCRIPTION
**Pull request by chipaca**
This change makes it so that the snap ID is stored in the snapshot,
and if on restore that ID is non-empty, and there is a snap with the
snapshot's snap's name installed that has a non-empty snap ID, and
these two IDs are different, the restore is fails.

This feature was requested at the Brussels sprint, as a minimal
prerequisite to renames (and merges). This then needs to grow to
handle the rename, but at least for now it will do no damage.

### CODE REVIEW

File: overlord/snapshotstate/snapshotstate.go [hunk 1]
```diff
@@ -197,6 +202,30 @@ func Restore(st *state.State, setID uint64, snapNames []string, users []string)
 
 	ts = state.NewTaskSet()
 
+	for i, snapID := range snapIDs {
+		if snapID == "" {
+			// snapshotted snap was sideloaded, never mind
```
Comment (zyga) [inline overlord/snapshotstate/snapshotstate.go]: `// The snap referring to the snapshot was locally installed without assertions`

File: overlord/snapshotstate/snapshotstate.go [hunk 2]
```diff
@@ -197,6 +202,30 @@ func Restore(st *state.State, setID uint64, snapNames []string, users []string)
...
+			// snap not installed
+			continue
+		}
+		sideInfo := snapst.CurrentSideInfo()
+		if sideInfo == nil {
+			// ... shouldn't happen (but also not installed)
+			continue
+		}
+		if sideInfo.SnapID == "" {
+			// current snap is sideloaded, never mind then
```
Comment (zyga) [inline overlord/snapshotstate/snapshotstate.go]: Same re-wording suggestion as above.

File: overlord/snapshotstate/snapshotstate.go [hunk 3]
```diff
@@ -197,6 +202,30 @@ func Restore(st *state.State, setID uint64, snapNames []string, users []string)
...
+			continue
+		}
+		if sideInfo.SnapID != snapID {
+			return nil, nil, fmt.Errorf("cannot restore snapshot over id change: %.7s… → %.7s…", snapID, sideInfo.SnapID)
```
Comment (zyga) [inline overlord/snapshotstate/snapshotstate.go]: Will the -> arrow render OK on simple terminals? Perhaps call a helper to translate it to "->"

Comment (chipaca) [reply overlord/snapshotstate/snapshotstate.go]: It works on the linux console with default settings, which was the good-enough-for-now pass mark we've used before.

File: overlord/snapshotstate/snapshotstate.go (hunk 2 above)
Comment (chipaca) [reply overlord/snapshotstate/snapshotstate.go]: I'm going with s/sideloaded/unasserted/, thank you for pointing it out

**Pull request merged**